presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1354,11 +1356,26 @@ else if (expression instanceof LongLiteral) {
                     }
 
                     // otherwise, just use the expression as is
+                    boolean orderByOnExpression = false;
                     if (orderByExpression == null) {
                         orderByExpression = expression;
+                        orderByOnExpression = true;
                     }
 
                     ExpressionAnalysis expressionAnalysis = analyzeExpression(orderByExpression, sourceScope);
+                    if (orderByOnExpression) {
+                        // Try to resolve within orderByScope. If it is possible, then we warn - standard behavior differs from legacy one.
+                        for (Map.Entry<Expression, FieldId> columnExpression : expressionAnalysis.getColumnReferences().entrySet()) {
+                            Optional<ResolvedField> resolvedField = orderByScope.tryResolveField(columnExpression.getKey());
+                            if (resolvedField.isPresent()) {
+                                if (!FieldId.from(resolvedField.get()).equals(columnExpression.getValue())) {
```

Comment:
this will give false warnings, for instance:
```
SELECT 1 as a FROM (VALUES 1) t(a) ORDER BY (SELECT a FROM (VALUES 2) t(a));
```
This is because column references contains all references (including subquery). Like I mentioned previously, probably the cleanest approach is to separate name resolution from type resolution in `ExpressionAnalyzer` and do name resolution on order by expression using `orderByScope`